samples/go/csv/read.go

```diff
@@ -99,28 +100,64 @@ func ReadToList(r *csv.Reader, structName string, headers []string, kinds KindSl
 			panic(err)
 		}
 
-		fields := make(types.ValueSlice, len(headers))
-		for i, v := range row {
-			if i < len(headers) {
-				fieldOrigIndex := fieldOrder[i]
-				val, err := StringToValue(v, kindMap[fieldOrigIndex])
-				if err != nil {
-					d.Chk.Fail(fmt.Sprintf("Error parsing value for column '%s': %s", headers[i], err))
-				}
-				fields[fieldOrigIndex] = val
-			}
-		}
+		fields := readFieldsFromRow(row, headers, fieldOrder, kindMap)
 		valueChan <- types.NewStructWithType(t, fields)
 	}
 
 	return <-listChan, t
 }
 
+// getFieldIndexByHeaderName takes the collection of headers and the name to search for and returns the index of name within the headers or -1 if not found
+func getFieldIndexByHeaderName(headers []string, name string) int {
+	for i, header := range headers {
+		if header == name {
+			return i
+		}
+	}
+	return -1
+}
+
+// getPkIndices takes collection of primary keys as strings and determines if they are integers, if so then use those ints as the indices, otherwise it looks up the strings in the headers to find the indices; returning the collection of int indices representing the primary keys maintaining the order of strPks to the return collection
+func getPkIndices(strPks []string, headers []string) []int {
+	result := make([]int, len(strPks))
+	for i, pk := range strPks {
+		pkIdx, ok := strconv.Atoi(pk)
+		if ok == nil {
+			result[i] = pkIdx
+		} else {
+			result[i] = getFieldIndexByHeaderName(headers, pk)
+		}
+		if result[i] < 0 {
+			d.Chk.Fail(fmt.Sprintf("Invalid pk: %v", pk))
+		}
+	}
+	return result
+}
+
+func readFieldsFromRow(row []string, headers []string, fieldOrder []int, kindMap []types.NomsKind) types.ValueSlice {
+	fields := make(types.ValueSlice, len(headers))
+	for i, v := range row {
+		if i < len(headers) {
+			fieldOrigIndex := fieldOrder[i]
+			val, err := StringToValue(v, kindMap[fieldOrigIndex])
+			if err != nil {
+				d.Chk.Fail(fmt.Sprintf("Error parsing value for column '%s': %s", headers[i], err))
+			}
+			fields[fieldOrigIndex] = val
+		}
+	}
+	return fields
+}
+
 // ReadToMap takes a CSV reader and reads data into a typed Map of structs. Each row gets read into a struct named structName, described by headers. If the original data contained headers it is expected that the input reader has already read those and are pointing at the first data row.
 // If kinds is non-empty, it will be used to type the fields in the generated structs; otherwise, they will be left as string-fields.
-func ReadToMap(r *csv.Reader, structName string, headersRaw []string, pkIdx int, kinds KindSlice, vrw types.ValueReadWriter) types.Map {
+func ReadToMap(r *csv.Reader, structName string, headersRaw []string, primaryKeys []string, kinds KindSlice, vrw types.ValueReadWriter) types.Map {
 	t, fieldOrder, kindMap := MakeStructTypeFromHeaders(headersRaw, structName, kinds)
 
+	pkIndices := getPkIndices(primaryKeys, headersRaw)
+
+	topMap := types.NewMap()
```

Comment:
This variable is now being used to build the entire map in memory. I think doing it this way will break csv importer for large files. I've been thinking about how to build nested maps efficiently using our current streaming approach and, so far, I haven't hit upon any great ideas. Perhaps you can do something similar to what I did in nomdex_build.go. Create the map of keys as a go map of maps (presumably the keys are small enough that this can fit into memory though it's not a guarantee). At the lowest level, instead of adding rows to the map, create a streamingSet for each keys and insert rows into them. When you've finished reading all the rows, create streamingMaps from the bottom up and insert all entries in them.